sql/core/src/main/scala/org/apache/spark/sql/dynamicpruning/PartitionPruning.scala

```diff
@@ -0,0 +1,264 @@
+ * predicate on the partitioned table using the filter from the other side of
+ * the join and a custom wrapper called DynamicPruning.
+ *
+ * The basic mechanism for DPP inserts a duplicated subquery with the filter from the other side,
+ * when the following conditions are met:
+ *    (1) the table to prune is partitioned by the JOIN key
+ *    (2) the join operation is one of the following types: INNER, LEFT SEMI (partitioned on left),
+ *    LEFT OUTER (partitioned on right), or RIGHT OUTER (partitioned on left)
+ *
+ * In order to enable partition pruning directly in broadcasts, we use a custom DynamicPruning
+ * clause that incorporates the In clause with the subquery and the benefit estimation.
+ * During query planning, when the join type is known, we use the following mechanism:
+ *    (1) if the join is a broadcast hash join, we replace the duplicated subquery with the reused
+ *    results of the broadcast,
+ *    (2) else if the estimated benefit of partition pruning outweighs the overhead of running the
+ *    subquery query twice, we keep the duplicated subquery
+ *    (3) otherwise, we drop the subquery.
+ */
+object PartitionPruning extends Rule[LogicalPlan] with PredicateHelper {
+
+  /**
+   * Search the partitioned table scan for a given partition column in a logical plan
+   */
+  def getPartitionTableScan(a: Expression, plan: LogicalPlan): Option[LogicalRelation] = {
+    val srcInfo: Option[(Expression, LogicalPlan)] = findExpressionAndTrackLineageDown(a, plan)
+    srcInfo.flatMap {
+      case (resExp, l: LogicalRelation) =>
+        l.relation match {
+          case fs: HadoopFsRelation =>
+            val partitionColumns = AttributeSet(
+              l.resolve(fs.partitionSchema, fs.sparkSession.sessionState.analyzer.resolver))
+            if (resExp.references.subsetOf(partitionColumns)) {
+              return Some(l)
+            } else {
+              None
+            }
+          case _ => None
+        }
+      case _ => None
+    }
+  }
+
+  /**
+   * Insert a dynamic partition pruning predicate on the left side of the join using the filter
+   * on the right side of the join.
+   *  - to be able to identify this filter during query planning, we use a custom
+   *    DynamicPruning expression that wraps a regular In expression
+   *  - we also insert a flag that indicates if the subquery duplication is worthwhile and it
+   *  should run irrespective the type of join, or is too expensive and it should be run only if
+   *  we can reuse the results of a broadcast
+   */
+  private def insertPredicate(
+      pruningKey: Expression,
+      pruningPlan: LogicalPlan,
+      filteringKey: Expression,
+      filteringPlan: LogicalPlan,
+      joinKeys: Seq[Expression],
+      hasBenefit: Boolean,
+      broadcastHint: Boolean): LogicalPlan = {
+    val reuseEnabled = SQLConf.get.dynamicPruningReuseBroadcast
+    val index = joinKeys.indexOf(filteringKey)
+    if (hasBenefit || reuseEnabled) {
```

Comment:
hmm, when would users disable `dynamicPruningReuseBroadcast`?

Reply:
It's just a fallback option and it's an internal conf.